### PR TITLE
#15572: Rewrite of Reshape OP from scratch

### PR DESCRIPTION
**Pull request by jvegaTT**
### Ticket
Solves the following issues (and potentially others, still going through and testing opened issues):
https://github.com/tenstorrent/tt-metal/issues/13889
https://github.com/tenstorrent/tt-metal/issues/12153
https://github.com/tenstorrent/tt-metal/issues/14676
https://github.com/tenstorrent/tt-metal/issues/15048
https://github.com/tenstorrent/tt-metal/issues/15269
https://github.com/tenstorrent/tt-metal/issues/14515
https://github.com/tenstorrent/tt-metal/issues/14513
https://github.com/tenstorrent/tt-metal/issues/14972
https://github.com/tenstorrent/tt-metal/issues/15524

### Problem description
Reshape OP was re-written from scratch and has no host fallback during legal uses

Previous issues were: 

A lack of higher dimensionality support. I built a dimensionality reducer that found that for any arbitrary ND shape there exists a 3D shape in tile or a 2D shape in row major where conversion from this shape to the original does not require data movement (metadata only).  We now do a ND->MD reshape by doing ND->3D->reshape->3D->MD in tiled and ND->2D->reshape->2D->MD in row major where we select the shapes of the two 2/3D intermediate shapes such that ND->2/3D and 2/3D->MD are metadata only changes. This allows us to only have to support 3D->3D shapes in tile and 2D->2D in row major in the kernel and get generality for free

A PCC issue which was caused by an invalid optimization which would produce a metadata only reshape even when some of the data should have been moved. In this case the data corruption was small but not 0 so it often ended up hidden in most invalid calls. This has been removed which hurt perf a bit but it is a necessary change.

The reshape function has been generalized to work in every case removing the need for reshape on host fallback for valid cases.

The old implementation unnecessarily needed a CB large enough to hold the entire tensor. New version only needs to allocate CBs whose total area is page_size_of_input + page_size_of_output + a small constant. This enables the reshape of very large tensors that almost fully occupy the L1. This also allows us to parallelize reading and writing by only adding barriers when needed rather than a big barrier between the reader and the writer kernels. 

### What's changed
Complete re-write of reshape OP - Sweep pass rate is 98.57%

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/12164049672
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/12166706962 https://github.com/tenstorrent/tt-metal/actions/runs/12157327133/job/33903077970
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by jvegaTT** [comment]
Sweep tests are showing 98.57% on this branch for reshape

File: ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp
Comment (ntarafdar) [inline ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp:20]: This is great. Can you announce this utility function in TT developers. This might be useful for others.

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape_pybind.cpp
Comment (ntarafdar) [inline ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape_pybind.cpp:47]: Amazing! Is therr a condition for 0 cost view that this is left as null, because in the case where we want to go from L1 to dram or vice versa or sharded to interleaved or vice versa it will require a data movement

**Review by ntarafdar — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape_pybind.cpp
Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape_pybind.cpp:47]: Oh true, just added it

**Comment by jvegaTT** [comment]
Update to those waiting on this PR. This generality push while in the average case has been found to improve in perf (no host fallback!), some perf tests have slightly diminished perf compared to the previous implementation. This is partially due to the fact some of the bug fixes unavoidably increase compute time and thus I hope the other enhancements will compensate. I have a group of optimizations I have envisioned to be applied, one was pushed last week which implements this in multiple cores that brought us very close to passing all tests. Hopefully will be able to reach the targets when all have been implemented.

**Comment by jvegaTT** [comment]
In particular the biggest culprit is fixing the PCC errors. We were in an unsafe manner doing only metadata changes when only a small, but not zero, amount of data movement was required. This means it was a coin flip if enough data corruption would occur to trigger a PCC assertion or not. However removing this super unsafe error resulting in corruption means some functions went from a 0 cost metadata change to an on-device reshape which is more expensive (not as bad as in-host but still pretty bad).

**Comment by jvegaTT** [comment]
I believe the perf issue is really a non-issue. I have checked the code of the device perf test and it looks like it uses tracy to record only the device execution times without consideration for host fallback and synchronization delays. As we have moved reshape to being always on device, in the old code, the times when it would rely on host fallback would have recorded a device execution time of 0ns even though the model overall took much much longer to run. I have reduced the expected device times to account for this.

**Review by sjameelTT — COMMENTED** [review]
What are the remaining failures from?

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/device/rm_reshape_interleaved_writer.cpp
Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/device/rm_reshape_interleaved_writer.cpp:6]: Yes it is unfortunately, it was when I was trying to split it to run in two risk cpus in each core. However it had a bug and then I thought of an optimization that cuts memory copying in half but makes it that you can't split the work between risks so the benefit disappeared. I think I should save an archive of the old code attempting that but delete it from this PR.

Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/device/rm_reshape_interleaved_writer.cpp:6]: This entire function is never run though so I can just gut it out

**Review by rfurko-tt — COMMENTED** [review]
Hi @jvegaTT !
Thanks for the great work!

Could you please explain what does "legal use" mean? Thanks in advance!
`Reshape OP was re-written from scratch and has no host fallback during legal uses`

**Comment by jvegaTT** [comment]
It means in every case that torch would also support, that is that the logical volume does not change

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/device/rm_reshape_interleaved_writer.cpp
Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/device/rm_reshape_interleaved_writer.cpp]: I deleted the two unused files (and put them on a different branch in case we decide to use them down the road)

**Comment by jvegaTT** [comment]
To highlight the prevalence of host fallback, according to sweeps 11% of all sweep tests from the models team required fallback on host in the implementation currently on main. This is after the previously merged view operation eliminated many of them.

**Review by yan-zaretskiy — APPROVED** [review]

File: tests/ttnn/unit_tests/test_reshape.py
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/test_reshape.py:316]: this works now, right? "test_broken_reshape" is such a confusing name :D

**Review by TT-BrianLiu — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape_pybind.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape_pybind.cpp:58]: how do you choose between 0 and nullptr here?

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape_common.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape_common.hpp:5]:
```
using PadValue = std::variant<uint32_t, float> PadValue;
```

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.hpp:38]: want to doublecheck. I did not see `launch_op`, should this be `register_operation_with_auto_launch_op`?

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:315]: please avoid c-style casts. prefer c++ casts like static_cast, dynamic_cast, reinterpret_cast

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:314]: looks like clang format was not run here

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:314]: @blozano-tt 👀

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:306]: How soon will we actually have a view and reshape call so its really clear when cost is guaranteed to be ~0

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/host/reshape_rm_host_prep.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/host/reshape_rm_host_prep.cpp:146]: move to a separate method?

**Review by ayerofieiev-tt — APPROVED** [review]

File: tests/ttnn/unit_tests/test_reshape.py
Comment (jvegaTT) [reply tests/ttnn/unit_tests/test_reshape.py:316]: I just copied the code from the issue, I am going to rename the test to test_previously_failing_test

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape_pybind.cpp
Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape_pybind.cpp:58]: I set the default directly when it is an integer and there is no type ambiguity. In the case of pad_value it actually accepts both float or int and interprets it based on the datatype of the tensor. So if it is a bfloat16 tensor, the default is 0.0 but in a uint32 tensor the default is integer 0. In the case of memory_config I read in the configuration of the source tensor and match it to the output tensor if it is not specified.

File: ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp
Comment (tt-aho) [inline ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp:42]: Please do not add a new hardcoded define for max_packet_size. We already have `NOC_MAX_BURST_SIZE`.

We also already updated some apis to do this sort of thing, ex noc_async_write is already templated to support using the one packet api if you specify a size less than the max transfer size. You could instead copy how noc_async_write was templated for noc_async_read.

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.hpp
Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.hpp:38]: I have operation::run in reshape.cpp line 246. I don't always run the device part of the operation since metadata only changes do not go on device, I want to be able to control if the operation will be run or not. Maybe my understanding of how operation registering is wrong?

File: ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp
Comment (tt-aho) [inline ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp:44]: A number of typos alligned

File: ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp
Comment (tt-aho) [inline ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp:48]: You should also not be hardcoding alignment values like 16 and 64 since there's no guarantee they're always that value across arches, these can be queried and passed from host as compile time args or accessed in kernels ex L1_ALIGNMENT.

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp
Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:306]: There is a separate PR right now for metadata-only reshapes that will cause PCC issues if incorrectly used. However I like to consolidate the guidance to always use ttnn.reshape that does the view if safe to do so. Maybe I can add a can_do_view operation which will tell the model writer if it is going to be a view or not in advance? I have also provided clear guidance in the pybind to state when a 0 cost view is happening.

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/host/reshape_rm_host_prep.cpp
Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/host/reshape_rm_host_prep.cpp:146]: Almost every OP does it this way though, I wanted to maintain consistency.

File: ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp
Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp:42]: It has the max packet size support but not the only_writes support which gives  us just a bit more perf as some of the data mover ops only have reader or writer in a kernel. I think we should keep it as is but will migrate when the only writes optimization is also there. I am changing the max page size to NOC_MAX_BURST_SIZE though.

File: ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp
Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp:44]: Lol, English not being my first language is showing

File: ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp
Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp:48]: This is a bit annoying to fix as I am also using MASK and OFFSET that is hardcoded to what should be used if the shift 16 and 64 respectively and they are also used in other files as this is the common code for all data mover ops. I can't take the value as compile time arg as this is in common. However because it is in common I think updating it should a new arch use a new alignment is not very difficult.

File: ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp
Comment (tt-aho) [reply ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp:42]: We didn't expose only read/only write flags in the original noc apis because initial testing seemed to show it was mostly functional, but there were some kernels where it seemed to hang even though it should have only been issuing only reads/writes so it was never merged since it risks users running into issues. It's fine if you want to use/expose it outside of dataflow api but you should be careful about exposing a common api for users that hasn't been fully tested. Ex. GS does not support > 8K txns at all, but that is not handled correctly here to ignore the flag for GS.

My suggestion here is to remove you logic regarding max transfer size here since it is already handled in noc_async_write, and instead pass the value down.

Ex
```
template <uint32_t max_transfer_size, bool only_writes>
FORCE_INLINE void enhanced_noc_async_write(
    const uint32_t src_l1_addr, const uint64_t dst_noc_addr, const uint32_t bytes) {
    // If you do not know the max_transfer_size at compile time write 0 to it.
    // only writes is true if we ONLY use noc_async_read and all calls to tt_memmove have use_read_datamover as False
    if constexpr (only_writes) {
        noc_async_write_one_packet(src_l1_addr, dst_noc_addr, bytes);
    } else {
        noc_async_write<max_transfer_size>(src_l1_addr, dst_noc_addr, bytes);
    }
}
```

File: ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp
Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp:48]: Come to think of it, I think the cleanest fix would be to define a DDR_ALIGNMENT at the same location L1_ALIGNMENT is defined that is set automatically at compile time depending on the ARCH type. The OPs would then use that constant value instead which is auto updated to match the architecture. However that would not be the solution for this PR

Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp:48]: Would it be acceptable to let this merge in as is for now and then I will post an issue where I can fix it in a cleaner way, just in the interest of getting this in as soon as possible.

Comment (tt-aho) [reply ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp]: DRAM_ALIGNMENT is also already defined (is this what you meant by DDR_ALIGNMENT?). It's fine if you want to address later, I'm not blocking the pr on any of these comments but wanted to make sure this was considered, as there is currently work being done regarding what alignments are being used.

File: models/demos/bert_tiny/tests/test_performance.py
Comment (davorchap) [inline models/demos/bert_tiny/tests/test_performance.py:122]: is this a perf drop or increase?

File: ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp
Comment (jvegaTT) [reply ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp]: Oh cool, good to know. Yeah will do a subsequent PR since I do want to avoid having to re-visit alignment when the next arch comes in

File: models/demos/bert_tiny/tests/test_performance.py
Comment (jvegaTT) [reply models/demos/bert_tiny/tests/test_performance.py:122]: Perf increased but the nature of the test makes it report a decrease. The device test currently uses tracy to record how long it runs on the device during the duration of the model. However, we are currently making the really expensive on-host reshapes run on device instead which has a significant end-to-end perf improvement. However this also means device run-time goes up since the device is doing more and the host is doing less. That is why we also ran full model experiments and saw that we significantly improved perf when taken in a wholistic approach. We do however need to update these targets to reflect the new device runtime though.

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/host/reshape_rm_host_prep.cpp
Comment (sjameelTT) [reply ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/host/reshape_rm_host_prep.cpp:146]: That style is deprecated now. For existing ops it's fine, but we do prefer this variant:

https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/adding_new_ttnn_operation.html

**Review by uaydonat — APPROVED** [review]

File: models/demos/bert_tiny/tests/test_performance.py
Comment (davorchap) [reply models/demos/bert_tiny/tests/test_performance.py:122]: freaking excellent, great job @jvegaTT !

**Comment by davorchap** [comment]
37 commits :) , squash to 1 commit?

e.g., 
"#0 : beast mode re-shape re-write for perf improvements and 98.57% pass rate"

**Comment by jvegaTT** [comment]
> 37 commits :) , squash to 1 commit?
> 
> e.g., "#issue : beast mode re-shape re-write for perf improvements and 98.57% pass rate"

Yup, doing a squash and merge now

**Comment by ntarafdar** [comment]
> 37 commits :) , squash to 1 commit?
> 
> 
> 
> e.g., 
> 
> "#issue : beast mode re-shape re-write for perf improvements and 98.57% pass rate" 

This would be an amazing commit message

**Pull request merged**
